### PR TITLE
Migrate interactive app to JSONForms

### PR DESCRIPTION
**Pull request by nx10**
chrome_9LlDHweCr7
chrome_AtleN6T0Z3
image

This PR migrates the interactive app to [JSONForms](https://jsonforms.io/).

- Much less code over-all, as the UI is auto-generated)
- Adapt JSON schema
- JSON data export & import
- CSV data export 
- List of Links/URLs can now be added directly
- Styles (of the app and shield icons) now match the [NMIND Proceedings website ](https://www.nmind.org/proceedings)

### CODE REVIEW

File: app/public/manifest.json
Comment (shnizzedy) [inline app/public/manifest.json]: I don't really care if these values change, but I don't think we intend to make them _more_ generic

**Comment by effigies** [comment]
```
Starting the development server...

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:69:19)
    at Object.createHash (node:crypto:133:10)
    at module.exports (/home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/util/createHash.js:90:53)
    at NormalModule._initBuildHash (/home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/NormalModule.js:401:16)
    at handleParseError (/home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/NormalModule.js:449:10)
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/NormalModule.js:481:5
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/NormalModule.js:342:12
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:373:3
    at iterateNormalLoaders (/home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:214:10)
    at iterateNormalLoaders (/home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:221:10)
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:236:3
    at runSyncOrAsync (/home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:130:11)
    at iterateNormalLoaders (/home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:232:2)
    at Array.<anonymous> (/home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:205:4)
    at Storage.finished (/home/chris/Projects/nmind/standards-checklist/app/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:55:16)
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:91:9
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=3.2.1 <3.5.0

YOUR TYPESCRIPT VERSION: 4.9.5

Please only submit bug reports when using the officially supported version.

=============
/home/chris/Projects/nmind/standards-checklist/app/node_modules/react-scripts/scripts/start.js:19
  throw err;
  ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:69:19)
    at Object.createHash (node:crypto:133:10)
    at module.exports (/home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/util/createHash.js:90:53)
    at NormalModule._initBuildHash (/home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/NormalModule.js:401:16)
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/NormalModule.js:433:10
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/webpack/lib/NormalModule.js:308:13
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:367:11
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:233:18
    at context.callback (/home/chris/Projects/nmind/standards-checklist/app/node_modules/loader-runner/lib/LoaderRunner.js:111:13)
    at /home/chris/Projects/nmind/standards-checklist/app/node_modules/babel-loader/lib/index.js:51:103
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}

Node.js v18.17.1
```

File: app/src/schema.json
Comment (shnizzedy) [inline app/src/schema.json]: :grimacing:
```suggestion
  "description": "Checklists for bronze, silver, and gold tiers of NMIND standards",
```

File: app/src/schema.json
Comment (shnizzedy) [inline app/src/schema.json:45]: I don't love using integer keys here, because
* the key doesn't have semantic relevance to the item
* I think the sequence of these items within a tier is arbitrary, but numbering them implies otherwise
* as the checklist evolves, we'd either end up with a sequence of numbers with gaps or reusing keys to mean different things across versions

File: app/src/schema.json
Comment (shnizzedy) [inline app/src/schema.json:46]: Do we need one more level of depth for the items, to include the boolean and [an optional comment](https://github.com/nmind/standards-checklist/issues/18)?

File: app/src/schema.json
Comment (shnizzedy) [inline app/src/schema.json:127]: For some items, like this one, where the answer could be "not applicable", I think we'll need
```JSON
"anyOf": [
  { "type": "boolean" },
  { "type": "string", "enum": ["not applicable"] }
]
```

File: app/src/checklist-schema.json
Comment (shnizzedy) [inline app/src/checklist-schema.json:1]: I think this file is vestigial, replaced by [`app/src/schema.json`](app/src/schema.json)?

File: app/src/utils.ts
Comment (shnizzedy) [inline app/src/utils.ts:146]: Once handling <span title="'not applicable's">'n/a's</span> is addressed, we'll need to discount those items from the total.

File: app/public/index.html
Comment (shnizzedy) [inline app/public/index.html]: I don't really care if we drop the license notice comment, but I think it's out of scope for this PR

File: app/README.md
Comment (shnizzedy) [inline app/README.md:1]: Moving from `interactive` to `app` makes sense to me since I don't think we're humoring the idea of a static version anymore, but:
* Is this still a [CRA](https://github.com/facebook/create-react-app) project directory?
* Either way, do we want / need this `app/README.md`, or should we move the how-to stuff to the top-level README?

File: app/src/App.test.tsx
Comment (shnizzedy) [inline app/src/App.test.tsx]: :frowning:

File: interactive/src/nmind.orgExcerpts.css
Comment (shnizzedy) [inline interactive/src/nmind.orgExcerpts.css:1]: I like the clarity of having a separate cascading style sheet that's specifically style from the main site, but I'm happy to be voted down about that

File: app/src/schema.json
Comment (shnizzedy) [inline app/src/schema.json:1]: ### I. This file vs. `checklist.json`

Is the idea to move away from the [JSON-LD encoding](https://www.repronim.org/reproschema/) in [`checklists/checklist.json`](https://github.com/nmind/standards-checklist/blob/f3418d9b8db4aeed1be441ea700b17b40dfa4f42/checklists/checklist.json) or to have it also encoded in JSON schema?

If the former, I think we can replace `checklist.json` with this JSON.

If the latter, I think we should choose one encoding to maintain and write a CI or precommit script to handle the translation.

### II. Checklist versioning

The current directory structure is with the idea that the checklist content will be versioned separately from the checklist interface. I think that's the right design, but I don't know how to do that without housing the schema in a sibling directory tree with its own `package.json`.

Depending on what the answers are to part I of this comment, this part could be moot if the JSON-LD sibling checklist is the source of truth that is translated into the JSON Schema checklist for the interface.

**Review by shnizzedy — APPROVED** [review]
Looks good to me! I'd maybe delete `app/src/checklist-schema.json` before merging (if that file's really unnecessary like it seems to me), but otherwise, I think all of my suggestions can become issues for future PRs

**Review by shnizzedy — DISMISSED** [review]
Changing from 'approve' to 'request changes' after seeing [Chris' comment](https://github.com/nmind/standards-checklist/pull/30#issuecomment-1749112354)

**Comment by shnizzedy** [comment]
I had to `npm ci --force` to get it to install (
```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: @jsonforms/material-renderers@3.1.0
npm ERR! Found: @mui/icons-material@5.14.9
npm ERR! node_modules/@mui/icons-material
npm ERR!   @mui/icons-material@"^5.14.9" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer @mui/icons-material@"~5.11.16" from @jsonforms/material-renderers@3.1.0
npm ERR! node_modules/@jsonforms/material-renderers
npm ERR!   @jsonforms/material-renderers@"^3.1.0" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: @mui/icons-material@5.11.16
npm ERR! node_modules/@mui/icons-material
npm ERR!   peer @mui/icons-material@"~5.11.16" from @jsonforms/material-renderers@3.1.0
npm ERR!   node_modules/@jsonforms/material-renderers
npm ERR!     @jsonforms/material-renderers@"^3.1.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution
```
), but it does work for me locally 
localhost:3000

File: app/public/manifest.json
Comment (nx10) [reply app/public/manifest.json]: Fixed with [81c5036](https://github.com/nmind/standards-checklist/pull/30/commits/81c5036fad5e9001829d9cb75a9b967ee5f6e847)

File: app/src/schema.json
Comment (nx10) [reply app/src/schema.json:45]: The keys are a direct migration from the previous e.g.:

https://github.com/nmind/standards-checklist/blob/f3418d9b8db4aeed1be441ea700b17b40dfa4f42/checklists/checklist.json#L70

Which is now `documentation.bronze.1`.

But I agree that these could be semi-informative short IDs instead.

File: app/src/schema.json
Comment (nx10) [reply app/src/schema.json:46]: Exactly, it would look something like this:

```JSON
"2": {
  "type": "object",
  "title": "Documentation is up to date with version of software",
  "properties": {
    "value": {
      "type": "boolean",
      "default": false
    }
    "comment": {
      "type": "string",
      "default": ""
    }
  }
},
```

To avoid repetition this could be [`$def`-ined once and then used for every item](https://json-schema.org/understanding-json-schema/structuring#dollardefs).

File: app/src/schema.json
Comment (nx10) [reply app/src/schema.json:127]: Makes sense!
Not applicable could also be a `boolean`

File: app/src/checklist-schema.json
Comment (nx10) [reply app/src/checklist-schema.json:1]: Good catch! React has so much boilerplate I did not see it. Fixed in [176b45a](https://github.com/nmind/standards-checklist/pull/30/commits/176b45a4b5b78f152bf855135edfca171b67d95d)

File: app/public/index.html
Comment (nx10) [reply app/public/index.html]: Copied it over in [02d7c65](https://github.com/nmind/standards-checklist/pull/30/commits/02d7c65a98cddacc5e9d1730ed0f136de168e5b8)

File: app/README.md
Comment (nx10) [reply app/README.md:1]: For clarification I did not move things over but started from "scratch" with the [JSONforms react seed repo](https://github.com/eclipsesource/jsonforms-react-seed). Git thinks I renamed the directory but it is actually completely new - Fortunately this helped find all the meta information standard texts you pointed out, which I did not realize existed in the boilerplate

File: app/src/App.test.tsx
Comment (nx10) [reply app/src/App.test.tsx]: Fixed the "test" in https://github.com/nmind/standards-checklist/pull/30/commits/d638a441c666ec5894b4ec57661c54132148ce51

**Comment by effigies** [comment]
Could you change `interactive` to `app` in the GHA workflow?

https://github.com/nmind/standards-checklist/blob/f3418d9b8db4aeed1be441ea700b17b40dfa4f42/.github/workflows/ghpages.yml#L24-L32

File: interactive/src/nmind.orgExcerpts.css
Comment (nx10) [reply interactive/src/nmind.orgExcerpts.css:1]: @gkiar suggested matching the [NMIND proceedings](https://www.nmind.org/proceedings), but I effectively just matched the fonts. Everything else is standard material design

File: .github/workflows/ghpages.yml
Comment (effigies) [inline .github/workflows/ghpages.yml:24]: Need to update L32, as well.

**Comment by gkiar** [comment]
@shnizzedy @effigies any chance you could re-review soon?

Thanks!

File: .github/workflows/ghpages.yml
Comment (shnizzedy) [reply .github/workflows/ghpages.yml:24]: Resolved in https://github.com/nmind/standards-checklist/pull/30/commits/2965d38617a5ee2b3545cd93170584dc43f5481b

**Comment by effigies** [comment]
Do we have a deployment somewhere to test? I tried pushing to my fork and this is what I get: https://effigies.github.io/standards-checklist/

File: app/README.md
Comment (shnizzedy) [reply app/README.md:1]: Resolved in https://github.com/nx10/standards-checklist/pull/1

**Comment by shnizzedy** [comment]
> Do we have a deployment somewhere to test? I tried pushing to my fork and this is what I get: https://effigies.github.io/standards-checklist/

https://github.com/nx10/standards-checklist/pull/1 should fix deployment

**Review by shnizzedy — COMMENTED** [review]
1. I didn't notice in my initial review, but these changes lose the `clarification`s from the UI. E.g, 
   | before | after |
   | --- | --- |
   | `bronze_doc_5` & `bronze_doc_6` before | `bronze_doc_5` & `bronze_doc_6` after |
2. I think re: "List of Links/URLs can now be added directly" / #23 (this PR should close #23 if merged) that we want to include a timestamp from when the review was completed and some record of who completed the review. These can be separate PRs, especially since I don't think we ever decided how we want to authenticate / credit reviewers.
3. This PR doesn't include UI for comments, but this PR should close #22 if merged (UI for comments can also be a separate PR)
4. I think we want the bronze sections to start out expanded for a new checklist and for sections to automatically collapse when completed and subsequent sections to expand (but #21 was opened over a year ago, so maybe we don't care? Either way, this PR should also close #21). The initial clicks to open the checklists seem like unnecessary user friction to me.
5. Do we want to address [this](https://github.com/nmind/standards-checklist/pull/30#discussion_r1348926965) in this PR or later?
   > Do we need one more level of depth for the items, to include the boolean and https://github.com/nmind/standards-checklist/issues/18?
   
   > Exactly, it would look something like this:
   >
   > ```JSON
   > "2": {
   >   "type": "object",
   >   "title": "Documentation is up to date with version of software",
   >   "properties": {
   >     "value": {
   >       "type": "boolean",
   >       "default": false
   >     }
   >     "comment": {
   >       "type": "string",
   >       "default": ""
   >     }
   >   }
   > },
   > ```
   >
   > To avoid repetition this could be [$def-ined once and then used for every item](https://json-schema.org/understanding-json-schema/structuring#dollardefs).
6. Do we want to address [this](https://github.com/nmind/standards-checklist/pull/30#discussion_r1348929921) in this PR or later?
   > For some items, like this one, where the answer could be "not applicable", I think we'll need
   > ```JSON
   > "anyOf": [
   >   { "type": "boolean" },
   >   { "type": "string", "enum": ["not applicable"] }
   > ]
   > ```

   > Makes sense!
   > Not applicable could also be a `boolean`
7. [These questions / comments](https://github.com/nmind/standards-checklist/pull/30/files#r1347759318) are still relevant
   1. > Is the idea to move away from the [JSON-LD encoding](https://www.repronim.org/reproschema/) in [checklists/checklist.json](https://github.com/nmind/standards-checklist/blob/f3418d9b8db4aeed1be441ea700b17b40dfa4f42/checklists/checklist.json) or to have it also encoded in JSON schema?
   2. > The current directory structure is with the idea that the checklist content will be versioned separately from the checklist interface. I think that's the right design, but I don't know how to do that without housing the schema in a sibling directory tree with its own `package.json`.

**Comment by nx10** [comment]
Regarding 1: I have added the `clarifications` to the JSONSchema and wrote a custom checkbox renderer to make them look just like they were before:

image

Regarding the other points I am waiting for judgement / decisions.

**Comment by gkiar** [comment]
@shnizzedy thanks for the thoughtful comments!

1. Looks like it's been taken care of
2. agree to kick the can
3. @nx10 am I understanding correctly that the schema now supports comments, but the UI does not? Would it be a lot to add that to the UI, or should we kick the can on this and do it in a follow-up PR? I'm happy to kick the can
4. I don't mind either way — I agree it seems like unnecessary friction, if trivial to address, let's, otherwise, let's open an issue and resolve downstream once this is live
5. This is related to 3, or am I missing something? happy to kick the can
6. Same as 5
7. i) I think we could move this towards JSON-LD while still being JSON schema, but for now we wanted to be able to build and validate a form from the schema, so that was the tug towards the latter standard. ii) good question, not sure either; can we agree to resolve/further decouple later?

Would love if we could get this "live" soon, and then there are certainly kinks that will need to be ironed out as we use it more!

**Comment by nx10** [comment]
3) This PR does not add user comment support for items in any way. It just reimplements the currently public version with jsonschema + jsonforms and minor additions.

Of course this is easy to add, but I assume opening separate PRs for this and Jon's other points would make them easier for you all to review.

**Comment by effigies** [comment]
I've poked around on a rendered site, and things overall look good, but haven't had the spare cycles to really do critique. If Jon and Greg are happy, I say let's merge.

**Comment by shnizzedy** [comment]
Yeah, I'm happy with merging and opening issues for the open threads here.

To answer a few questions / clarify a couple points:

> 5. This is related to 3, or am I missing something? happy to kick the can

3 is adding comment support to the UI, 5 is adding comment support to the schema

> 6. Same as 5

6 is adding "n/a" as a specific selectable option where relevant

> 7. i) I think we could move this towards JSON-LD while still being JSON schema, but for now we wanted to be able to build and validate a form from the schema, so that was the tug towards the latter standard. ii) good question, not sure either; can we agree to resolve/further decouple later?

The current schema is already JSON-LD

```mermaid
flowchart LR
LD["JSON-LD"] --> UI
```

; this PR introduces a JSON Schema document as a separate, parallel source of truth.

```mermaid
flowchart LR
LD["JSON-LD"]
Schema["JSON Schema"] --> UI
```

I'm suggesting we keep the JSON-LD as the source of truth and autogenerate the JSON Schema for the UI from the JSON-LD

```mermaid
flowchart LR
LD["JSON-LD"] --automation--> Schema["JSON Schema"] --> UI
```

All that said, I agree these are all issues to open / [cans to kick](http://www.hrwiki.org/w/images/7/76/hsnoplay.PNG) once this PR is merged in.

**Review by shnizzedy — APPROVED** [review]

**Review by mgxd — APPROVED** [review]
Haven't had the time to do a deep review, but overall this looks great, and fulfills the goal of actually being able to use the checklist while reviewing a project - I'm also in favor of kicking the can on the rest of the missing pieces. thanks @nx10 !!

File: app/src/schema.json
Comment (shnizzedy) [reply app/src/schema.json:45]: https://github.com/nmind/standards-checklist/issues/32

File: app/src/schema.json
Comment (shnizzedy) [reply app/src/schema.json:46]: https://github.com/nmind/standards-checklist/issues/33

File: app/src/schema.json
Comment (shnizzedy) [reply app/src/schema.json:127]: https://github.com/nmind/standards-checklist/issues/34

File: app/src/utils.ts
Comment (shnizzedy) [reply app/src/utils.ts:146]: https://github.com/nmind/standards-checklist/issues/34

File: app/src/schema.json
Comment (shnizzedy) [reply app/src/schema.json:1]: https://github.com/nmind/standards-checklist/issues/36

**Pull request merged**
